🐛 Removes MachineSet Controller owner for existing Machines

srm09 opened the pull request:
**What this PR does / why we need it**:
Checks if any existing machines have the MachineSet controllers as the owner and removes them in the reconcile phase since the Machine controller should be the owner

**Which issue(s) this PR fixes**:
Fixes #3145

ncdc commented:
/ok-to-test

vincepri commented on controllers/machine_controller_phases.go:
```diff
@@ -114,6 +114,18 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
 	}
 
+	// remove machineset controller onwer, if any since the Machine should be the owner.
+	controller := metav1.GetControllerOf(obj)
+	if controller != nil {
```
```suggestion
	if controller := metav1.GetControllerOf(obj); controller != nil {
```

srm09 replied:
Done

vincepri commented on util/util.go:
```diff
@@ -385,6 +385,13 @@ func ReplaceOwnerRef(ownerReferences []metav1.OwnerReference, source metav1.Obje
 	return ownerReferences
 }
 
+func RemoveOwnerRef(ownerReferences []metav1.OwnerReference, source metav1.OwnerReference) []metav1.OwnerReference {
```
Add godoc

srm09 replied:
Done

jzhoucliqr commented on controllers/machine_controller_phases.go:
```diff
@@ -114,6 +114,17 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
 	}
 
+	// remove machineset controller onwer, if any since the Machine should be the owner.
+	if controller := metav1.GetControllerOf(obj); controller != nil {
```
minor nit, but maybe check if controller.Kind == "MachineSet" ?

srm09 replied:
Doesn't the hardcoded value for ownerRef.Kind to `MachineSet` implicitly do the same thing?

jzhoucliqr replied:
Yeah, it does. If I understand correctly, for performance, without checking Kind it'll always go through all ownerreferences, no matter if the MacineSet is present or not.  Also, for readability, this might be a bit confusing, as we kinds of assume this controller is MachineSet, but in fact it maybe Machine.

srm09 replied:
I still feel adding the check for controller Kind, somehow duplicates the logic. I will defer to the better judgement of experienced folks like @vincepri

fabriziopandini replied:
My two cents.
I'm +1 to have the check on "controlled by a machineSet" more explicit for improving for readability.

For the same reason, might be worth to split the command in two-line so we will have `ownerRefs := util.RemoveOwnerRef(...)` which makes it clear we are removing from the list, and the `obj.SetOwnerReferences(ownerRefs)` in a separated line

vincepri replied:
It might be worth adding it for readability, + a comment explaining the reason why this has to be removed this way. In general, owner reference is a very finite set, performance-wise we wouldn't get many improvements.

fabriziopandini commented on util/util_test.go:
```diff
@@ -792,3 +792,45 @@ func TestIsSupportedVersionSkew(t *testing.T) {
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			newOwnerRefs := RemoveOwnerRef(ownerRefs, tt.toBeRemoved)
+			g.Expect(len(ownerRefs) != len(newOwnerRefs)).To(Equal(tt.wasRemoved))
```
What about using `HasOwnerRef` to check if `toBeRemoved` is gone?

srm09 replied:
Makes the test cleaner, good point

fabriziopandini commented:
LGTM pending implementation of the tests in machine_controller_phases_test.go

vincepri commented:
@srm09 code changes look good, we only need to get tests in place

sedefsavas commented on controllers/machine_controller_phases.go:
```diff
@@ -114,6 +114,19 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
+	// infra Machines, hence remove any existing machineset controller owner reference
+	if controller := metav1.GetControllerOf(obj); controller != nil && controller.Kind == "MachineSet" {
+		ownerRefs := util.RemoveOwnerRef(obj.GetOwnerReferences(), metav1.OwnerReference{
+			APIVersion: clusterv1.GroupVersion.Version,
```
This should be just `clusterv1.GroupVersion.String()`. 
ApiVersion has both group and version info: `cluster.x-k8s.io/v1alpha3`

srm09 replied:
Makes sense, done

vincepri commented:
@srm09 What's the status of this PR?

detiber commented on controllers/machine_controller_phases.go:
```diff
@@ -114,6 +114,19 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
+			APIVersion: clusterv1.GroupVersion.Version,
+			Kind:       "MachineSet",
+			Name:       controller.Name,
+			UID:        controller.UID,
+			Controller: pointer.BoolPtr(true),
```
These look to be ignored by the comparisons and should be able to be omitted
```suggestion
```

srm09 commented on controllers/machine_controller_phases_test.go:
```diff
@@ -752,6 +752,12 @@ func TestReconcileBootstrap(t *testing.T) {
+	It("updates the owner references for existing machines managed by Machine Sets", func() {
+
+	})
+})
```
Removed this and added the tests to existing suites.

srm09 commented:
/test pull-cluster-api-test

srm09 commented:
> @srm09 What's the status of this PR?

@vincepri I added the tests for the machine_controller_phases with the help of @sedefsavas and also incorporated the suggestions from @detiber

fabriziopandini commented:
@srm09 great work!
/lgtm

vincepri commented on controllers/suite_test.go:
```diff
@@ -116,3 +119,36 @@ var _ = AfterSuite(func() {
 	}
 })
+
+func ContainRefOfKind(kind string) types.GomegaMatcher {
```
Nice!

ncdc replied:
This should probably be ContainRefOfGroupKind and take in a group and a kind. (Also this is awesome, btw!)

ncdc commented on controllers/machine_controller_phases.go:
```diff
@@ -115,6 +115,17 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
 
+	// With the migration from v1alpha2 to v1alpha3, Machine controllers should be the owner for the
+	// infra Machines, hence remove any existing machineset controller owner reference
+	if controller := metav1.GetControllerOf(obj); controller != nil && controller.Kind == "MachineSet" {
```
It would probably be safest to check both controller.Kind and the group portion of controller.APIVersion

vincepri commented:
/milestone v0.3.7

sedefsavas commented on controllers/machine_controller_phases.go:
```diff
@@ -115,6 +116,23 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
+		}
+		if gv.Group == clusterv1.GroupVersion.Group {
+			ownerRefs := util.RemoveOwnerRef(obj.GetOwnerReferences(), metav1.OwnerReference{
+				APIVersion: clusterv1.GroupVersion.String(),
```
Crit. If previous MachineSet owner has v1alpha2 version, does this not fail to remove it?
Can we add this scenario to the tests if it is not there already?

srm09 replied:
We are just checking the Group portion of the APIVersion. I can add a separate test for the negative scenario here.

vincepri replied:
That's a good point, can we just pass in the `controller` variable in here instead of creating a new OwnerReference? Given that we're checking the Kind and Group now, it should be safe.

ncdc commented on controllers/suite_test.go:
```diff
@@ -116,3 +120,42 @@ var _ = AfterSuite(func() {
+}
+
+func (matcher *refKindMatcher) FailureMessage(actual interface{}) (message string) {
+	return fmt.Sprintf("Expected %+v to contain refs of Kind %s", actual, matcher.kind)
```
Please update to include the group

ncdc replied:
Bump

ncdc commented on controllers/suite_test.go:
```diff
@@ -116,3 +120,42 @@ var _ = AfterSuite(func() {
+}
+
+func (matcher *refKindMatcher) NegatedFailureMessage(actual interface{}) (message string) {
+	return fmt.Sprintf("Expected %+v not to contain refs of Kind %s", actual, matcher.kind)
```
Please update to include the group

ncdc replied:
Bump

ncdc commented on controllers/suite_test.go:
```diff
@@ -116,3 +120,42 @@ var _ = AfterSuite(func() {
+	}
+}
+
+type refKindMatcher struct {
```
Please rename to something like refGroupKindMatcher

ncdc replied:
Bump

srm09 replied:
I did an incorrect rebase and lost that commit, adding it back

ncdc commented on controllers/suite_test.go:
```diff
@@ -116,3 +120,42 @@ var _ = AfterSuite(func() {
+func (matcher *refKindMatcher) Match(actual interface{}) (success bool, err error) {
+	ownerRefs, ok := actual.([]metav1.OwnerReference)
+	if !ok {
+		return false, fmt.Errorf("ContainRefOfKind matcher expects a slice of OwnerReference")
```
```suggestion
		return false, errors.Errorf("expected []metav1.OwnerReference; got %T", actual)
```

vincepri commented:
@srm09 Do you have time to address the comments above?

srm09 commented:
@vincepri I have resolved all the review comments.

ncdc commented on controllers/machine_controller_phases_test.go:
```diff
@@ -681,13 +681,22 @@ func TestReconcileBootstrap(t *testing.T) {
 		},
 		{
-			name: "existing machine, bootstrap provider is not ready",
+			name: "existing machine, bootstrap provider is not ready, and ownerref updated",
```
Sorry to think of this at the last minute, but could you please add a **new** test case that hard-codes the v1alpha2 API version for a MachineSet owner, just to cover all our bases? Thanks!

ncdc commented:
@srm09 just a couple of small issues to resolve and then we're good to merge. Thanks for your patience with the reviews!

fabriziopandini commented:
/lgtm
